### DISCUSSION TITLE
NES Language Frontend and Optimizer

### DISCUSSION THREAD

**Discussion by klauck** [nebulastream/nebulastream#47] [Ideas]
Many open-source projects offer reusable components (e.g., language frontends, intermediate plan representations, query optimizers, execution engines) for data management systems. (see paper "The Composable Data Management System Manifesto").

Proposed components differ in scope, maturity, complexity, extensibility, and (programming language) interface.

TODO: List advantages and disadvantages of project reuse (see paper "The Composable Data Management System Manifesto")

In the following, we list and abstract potential frontend and optimizer projects to use in NES. (@klauck: Refinement and extension in progress.)

**Frontend projects:**

Transform user queries into an intermediate representation that can be optimized and executed

- SQL-based:
usually use grammar and tokenization libraries, such as Flex, Bison or Antlr
    - Google's ZetaSQL
    - Meta's CoreSQL
    - PostgreSQL parser
    - Hyrise parser
    - Calcite (includes optimizer)
    - DataFusion (focus on execution engine)
- Non-SQL APIs
    - Calcite
    -  Ibis
    - ...

**Intermediate representations**
- all optimizers, execution engines and language frontends have to use a IR as input or output
- Substrait
  - focus on IR
  - tries to provide a unified and extensible IR specification

**Optimizers (projects)**
- Orca:
    - paper "Orca: A Modular Query Optimizer Architecture for Big Data"
    - Greenplum/Pivotal
    - C/C++
    - XML-based exchange with execution engine
    - "hard to integrate with non-PostgreSQL systems" (read Integrating the Orca Optimizer into MySQL)
    - potentially not very active open-source project
- Calcite
    - paper "Apache Calcite: A Foundational Framework for Optimized Query Processing Over Heterogeneous Data Sources"
    - Used by Flink, Hive, and other Apache Java projects
    - IR translation and integration (e.g., statistics) may be difficult for non-Java systems
    - includes language frontend and IR
- DataFusion
   - paper "Apache Arrow DataFusion: a Fast, Embeddable, Modular Analytic Query Engine"
   - Primarily an execution engine but includes frontends, IRs and optimizer
   - Rust
   - recent (trending?)
- Hyrise
   - paper "Hyrise Re-engineered: An Extensible Database System for Research in Relational In-Memory Data Management"
   - C++
   - no separate project
- other DBMS, e.g., DuckDB, PostgreSQL 
     

see also Firebolt paper "Assembling a Query Engine From Spare Parts"

**Comment by PhilippGrulich** [comment]
One general aspect to consider is that these components are usually quite large, require third party dependencies, and have a certain complexity on their own.

**Reply by zeuchste** [reply]
Yes, and another point is to consider is that we are not sticking together a product for a company here rather than we build systems to also do research and things differently. Also, we are bound to their schedule and their APIs and then we basically run after n different projects with are not in sync. Again for a product, this makes maybe sense to get to a solution fast but we want to do groundbreaking and disruptive research this means we cannot only rely on them  From my point of view we should go like this:

### Parser
Definitely, we should use one existing one as nobody in our group does parsing as a research topic, the question is how extensible they are as we want to introduce new concepts etc, and do not want to stick with plain SQL. The most important thing here is that it supports what we plan to do with the new algebra and is easy to use and understand and catches many of the errors in a good way for free for us.

### Optimizer
Here it gets tricky, what is an optimizer? also if we choose a parser from framework x and want to have an optimizer it would make sense to stay in the same framework as calcite. 

1. Rewrite Rule Engine: We need something to do logical rewrite rules, this we have currently in NES but this would also be supported by most optimizer frameworks
2. Cost-Based Plan Choosing: Based on a plan enumerator we choose for one query the best plan based on a cost model. We do not have this. The cost anyhow has to come from NES itself so it needs to be injectable into the framework (I know Calcite is to some extent capable of it). Then the Enumerator needs to create equivalent plans and estimate them. I think this is part of most optimizer frameworks and could be used by NES as NES does not have it
3. Multi-Query Optimization: As far as I know no framework supports this they are all designed to take one query and optimize it. They do not support a running plan and additions. 
4. Placement: As far as I know no framework supports this, yes it could be hacked into calcite by node properties but those frameworks are not built for that
5. Deployment: As far as I know no framework supports the deployment and maintenance of queries they are all one-shot approaches
6. Continuous re-optimization: We do not fire and forget we need to have the possibility to re-optimize and change based on query and infrastructure changes. this is not supported at all

My personal takeaway with my current knowledge is that we could solve 1 and 2 with a framework but for the rest we have to stick with what we have, this is not a problem and would really simplify the tasks of our coordinator but we cannot replace it.

We also have to take into account that we want in the future a multi-optimizer setup with region optimizer that works together and we should not exclude this by design.

### Engine
Yes maybe Velox could replace our engine but why? yes, we should better find out the differences but again if NES is a set of existing systems then the only novelty is the glue code. But yes we should look into this

### IR
So we do have our NES IR with a lower level IR. I still do not really get Substrate if we only have one possible optimizer as it is basically just a JSON representation of a query. If you want to have different optimizers yes maybe this might change and Substrate is as far as we say not really mature. We really need to dive deeper here with the pros and cons compared to the internal format we are currently using.

**Comment by PhilippGrulich** [comment]
Another interesting project here is arrow.
It provides a in memory layout and sources/sinks for some data formats. Using arrow could help us to get the null support correctly. Furthermore it provides some components for memory management

**Comment by ankitgit** [comment]
I am writing some of my thoughts and suggestions regarding this topic. 

## Frontend/Query API:

Anltlr will easily integrate into our system to support full SQL grammar. With a bit of effort, you can also extend it to support any new grammar you might define as part of the streaming algebra you want to develop. However, I suggest using our logical plans or developing something new for any intermediate representation if you observe any limitations. This will give you full flexibility to change/extend it later. Any 3pp IR that you might want to use will require integrating it with the system (which can be challenging), might force you to use their optimizer, or won't support any new construct that you might want to use. 

## Optimizer:

We should check these libraries for their support for optimizing window-based queries, as these are the dominant constructs in stream processing systems. When I checked, these libraries didn't support optimizing windowed queries (Calcite seemed only to provide grammar support), as they are primarily designed for optimizing queries that need execution on the databases (please let me know if this is not true). Additionally, none of these libraries extend the support for operator placement optimization, multi-query optimization, infrastructure-specific optimizations, or handling of workload or infrastructure dynamism. So, we should consider our existing optimizer's features before migrating to a 3PP optimizer with low extensibility and difficult integration.

Nonetheless, we can still use these libraries to apply heuristics-based rewrite rules to our queries involving standard relational operators (filer, projection, etc.). For this, I suggest using the Postgres optimizer, which is written in C++ and might be easy to integrate. DuckDB also used it initially, but they later wrote their own rewrite rules (https://github.com/duckdb/duckdb/tree/main/src/optimizer), and I guess it is not used anymore. Our optimizer has also defined a collection of cool rewrite rules as well.  We should define further rewrite rules that leverage underlying infrastructure and other system properties. For example, the distribution or physical partitioned sources can enable us to optimize our queries further, which the Sota systems do not consider otherwise.

**Reply by zeuchste** [reply]
**To Frontend:** My biggest fear would be that we have to to the error handling and messaging for every weird edge case by our own and my hope was that Calcite or others made a lot of efforts there to properly error message and check all the existing edge cases. If this is not the case, I think there is no meaning in using them. One other thing would be if we could borrow their grammar and thus the same syntax as they do to make us compatible

**For the optimizer:** I totally agree with this and I do not see with my current overview anything that could replace what we have. As I said above, the only thing that we could borrow is the parser and the rule engine. For the rule engine we should spent time in comparing what we have with what they have and maybe combine them or find pros and cons at lest.

For example this is the out-of-the-box rule set that Calcine support, we should take a look which of those apply for us too and use them

[https://github.com/apache/calcite/tree/master/core/src/main/java/org/apache/calcite/rel/rules](link)

A good slide set of what is possible with Calcite is [https://percona.com/sites/default/files/presentations/Building%20Cost-Based%20Query%20Optimizers%20With%20Apache%20Calcite.pdf](this)

**Reply by klauck** [reply]
**To Frontend:** I agree that an existing project will support us with a good grammar, including edge cases. In this context, the parsers of products (duckdb and flink/calcite) are probably better than research parsers (e.g., NES and Hyrise). For Hyrise, for example, we have encountered some minor bugs over time.
Further, parsers of products come with ok-ish error messages for syntax errors: "ERROR:  syntax error at or near "FRM", but semantic checks must be provided or at least supported by us.  
I will, thus, look at the DuckDB parser, which is based on the PostgreSQL parser, first: https://github.com/duckdb/duckdb/tree/main/third_party/libpg_query

FYI: The PostgreSQL/DuckDB parser is based on Flex+Bison. (not on Antlr)

**Reply by zeuchste** [reply]
Hi but please node if we extend the grammar (which we plan to do) we also have to extend the error handling by our own

**Reply by ankitgit** [reply]
Btw for any kind of custom extensions in the grammar of Antlr or Flex+Bison, we always need to care for edge cases that can occur due to our extensions.Additionally, Antlr is currently integrated into our system for pattern specification language and I am working with antlr for supporting UDF equivalence identification. Therefore, if you can use Antlr it will be awesome as then we do not have to maintain different mechanisms.

**Reply by klauck** [reply]
Syntax errors will be handled by the parser (which is the hard part in my opinion).

**Reply by ankitgit** [reply]
Btw, regarding calcite: this page lists what stream features are not implemented in calcite https://calcite.apache.org/docs/stream.html#not-implemented

I am not sure if something has changed as this list is still on their official website.
